### ISSUE TITLE
Creating multiple domains with count fails due to spice port already in use

### ISSUE THREAD

**Issue by adamcstephens** [dmacvicar/terraform-provider-libvirt#84]
Terraform file:  https://gist.github.com/adamcstephens/7e56423ccceb3e1f3c0d808eb7cb8c59

(Ignore my serialconsole parameter, you'll be getting a PR for that. :)

```
Error applying plan:

3 error(s) occurred:

* libvirt_domain.centos.3: Error creating libvirt domain: [Code-1] [Domain-10] internal error: process exited while connecting to monitor: ((null):9070): Spice-Warning **: reds.c:2490:reds_init_socket: listen: Address already in use
2016-11-11T01:54:21.893813Z qemu-system-x86_64: failed to initialize spice server
* libvirt_domain.centos.0: Error creating libvirt domain: [Code-1] [Domain-10] internal error: process exited while connecting to monitor: annel/target/domain-3-centos0/org.qemu.guest_agent.0,server,nowait -device virtserialport,bus=virtio-serial0.0,nr=1,chardev=charchannel0,id=channel0,name=org.qemu.guest_agent.0 -spice port=5900,addr=127.0.0.1,disable-ticketing,seamless-migration=on -device cirrus-vga,id=video0,bus=pci.0,addr=0x2 -device virtio-balloon-pci,id=balloon0,bus=pci.0,addr=0x6 -object rng-random,id=objrng0,filename=/dev/random -device virtio-rng-pci,rng=objrng0,id=rng0,bus=pci.0,addr=0x7 -msg timestamp=on
Domain id=3 is tainted: high-privileges
((null):9075): Spice-Warning **: reds.c:2490:reds_init_socket: listen: Address already in use
2016-11-11T01:54:21.893813Z qemu-system-x86_64: failed to initialize spice server
* libvirt_domain.centos.2: Error creating libvirt domain: [Code-1] [Domain-10] internal error: process exited while connecting to monitor: x/store/qiic7ibvl75rrdmrq5d8w202165j25i9-dmidecode-3.0/sbin:/nix/store/bs6640nvs7wxn8qrg14snwzds7mzh7x9-dnsmasq-2.76/sbin:/nix/store/4pjlmmhm7cq9h53xbmm96y77xpibsk7b-ebtables-2.0.10-4/sbin:/nix/store/ldcxp98nkgnjzqs4ki7p7mxrhdgrzsjn-qemu-x86-only-2.7.0/sbin:/nix/store/8rn45r9ndfq5h7mx58r35p2szky5ja6n-coreutils-8.25/sbin:/nix/store/gjm2wjh8m72ch3ikiznn4v40h9mvpfks-findutils-4.6.0/sbin:/nix/store/h9aqgpgspgjhygj63wpncfzvz5ys851n-gnugrep-2.25/sbin:/nix/store/1awgi4ba6ymlzib2l7bxa13i17nq27d0-gnused-4.2.2/sbin:/nix/store/inbr0qixfmq0js9pwm8n0bhqad865mab-systemd-231/sbin QEMU_AUDIO_DRV=spice /nix/store/ldcxp98nkgnjzqs4ki7p7mxrhdgrzsjn-qemu-x86-only-2.7.0/bin/qemu-kvm -name guest=centos2,debug-threads=on -S -object secret,id=masterKey0,format=raw,file=/var/lib/libvirt/qemu/domain-1-centos2/master-key.aes -machine pc-i440fx-2.7,accel=kvm,usb=off -m 512 -realtime mlock=off -smp 1,sockets=1,cores=1,threads=1 -uuid 3c8941a6-37e6-45ad-a617-f5efab2188d4 -no-user-config -nodefa

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

**Comment by multani** [comment]
I tested the plugin on several machines and with a recent libvirt daemon running (4.0+), I don't have the problem anymore, whereas it's easily reproductible on 3.x.

I don't have access to them right now but there are several instances of the same problem reported on the libvirt bugtracker, even without Terraform involved, although I still didn't have understand what was the final fix to apply, beside doing a full upgrade of libvirt itself.

**Comment by danderson** [comment]
Interesting! I'll see if I can upgrade to a recent libvirt... Thanks for the tip!

**Comment by danderson** [comment]
Tried with libvirt 4.0, and sadly, still no luck :( Creating 4 VMs simultaneously, only 2 come up, the other 2 never start.

One option to work around this in the provider might be to add a lock just around VM creation, i.e. only allow 1 domain at a time to be in the state between the initial creation request, and the SSH connection establishing (or DHCP lease being issued). That would still let most of the costly provisioning happen in parallel, and just serialize the creation operations.

**Comment by multani** [comment]
That's weird, I think I managed to have 10 created at the same time on my desktop without any problem (then my machine was running out of memory and trashed, but that's another story.) I'll try to reproduce this and see if I can get you more information about my setup, maybe we'll be able to find the difference between the 2 configs.

**Comment by danderson** [comment]
For reference, my Terraform configs are at https://github.com/google/metallb/tree/master/terraform . It's fairly complex, I'm spinning up a VM on GCE with nested virtualization support, installing libvirt on that, and then bootstrapping 4 nested VMs.

If you have a GCP account and want to reproduce, you need to run `make-vmx-image.sh` first to create a Debian base image that has nested virt enabled. Then, remove the `depends_on` clauses in k8s_cluster.tf, and `terraform apply` should try to build everything.

In my experiments so far, I consistently get only 2/4 nested VMs created, the other two are in the "shutoff" state in `virsh list`. The exact VMs that get created vs. fail seems to change on each run.

**Comment by MalloZup** [comment]
https://github.com/dmacvicar/terraform-provider-libvirt/issues/84

**Comment by MalloZup** [comment]
I.could not reproducd this in new relesead version. Checked the code thi hs been fixed. Thx!

**Comment by Lennie** [comment]
Later I figured out what might have happened in my case, at some point I found a Qemu process still running even though libvirt said it was already gone.
